feat(http_server source): add all headers to the namespace metadata

sonnens opened the pull request:
This adds all the http headers to the source metadata if log_namespace is enabled

StephenWakely commented:
Thanks. It's not completely clear what the use case for this would be and it does add some extra processing overhead. 

I'm not seeing a relevant issue that this is solving. Would it be possible to raise one and add all the relevant details. We can discuss there if there is a need for this?

sonnens commented:
filed an issue with more details about why we wanted this

https://github.com/vectordotdev/vector/issues/18937

StephenWakely reviewed (dismissed):
We had a discussion around this change, and whilst we can see it would be useful for some - it is definitely not something that we want to have implemented all the time due to the overhead it introduces.

The way we would like to have this working is to change the [headers config](https://vector.dev/docs/reference/configuration/sources/http_server/#headers) to allow for wildcards. You can use the [glob](https://docs.rs/glob/latest/glob/) crate to do the [matching](https://docs.rs/glob/latest/glob/struct.Pattern.html#method.matches).

I'm not sure quite what the overhead will be of running the matching in `glob`. It may be worth detecting at load time if the  `header` contains any wildcards - if it does use `glob` otherwise use `==` like it currently does.

This way, if you want all the headers you can specify `headers: ["*"]`, and other users will be unaffected.

Possibly a little more involved that you initially anticipated, but this provides the best way forward for everyone. Let me know if you have any questions.

sonnens commented:
@StephenWakely Something like this look better?

neuronull commented on src/sources/http_server.rs:
```diff
@@ -330,8 +350,23 @@ impl SourceConfig for SimpleHttpConfig {
         let log_namespace = cx.log_namespace(self.log_namespace);
 
+        let (glob_headers, mut headers): (Vec<String>, Vec<String>) = self.headers
+            .clone()
+            .into_iter()
+            .partition(|s| s.contains('*'));
+
+        let glob_headers: Vec<glob::Pattern> = remove_duplicates(glob_headers, "headers")
+            .into_iter()
+            .map(|s| glob::Pattern::new(&s).unwrap())
+            .collect();
+
+        if !glob_headers.is_empty() {
+            headers = remove_glob_matches(&headers, &glob_headers, "headers");
+        }
+
         let source = SimpleHttpSource {
-            headers: remove_duplicates(self.headers.clone(), "headers"),
+            headers: remove_duplicates(headers, "headers"),
+            glob_headers: glob_headers.clone(),
```
👋  Thanks for adjusting the approach @sonnens 

The current design here of cloning the headers to create a separate list of glob headers is I think a different direction than what @StephenWakely had in mind and adds some storage and processing that should be avoidable.

I think the idea was to use the glob pattern matching in the `enrich_events()` function directly. i.e not having separate lists, but just using the `headers` config list as-is. Additionally, I think Stephens suggestion to check if the supplied headers contain any wildcard as a gate on using the glob pattern matching is a good idea to reduce overhead when wildcards are not used.

Some pseudo code that would go in enrich_events:

```
if any header in self.headers contains "*" {
  // use glob pattern matching to determine which headers to add to the event metadata
else {
  // use headers_config.get
}
```

Some other general notes:
- we should adjust the doc comment in `SimpleHttpConfig` for the `headers` member, to describe this new behavior. That doc comment is used to automatically generate our documentation on the website. 
- we should add a couple docs::examples for the wildcard case
- unit tests should be added to confirm the behavior works as expected.

sonnens replied:
My thought process for splitting in to headers/glob_headers is that HeaderMap.get is cheap but glob.match is much more expensive, and unless I'm misunderstanding the code (entirely probable) there's only one SimpleHttpConfig instantiated per source on startup, so it can eat a few clock cycles and a size_t there in order to save a bunch on every ingest event.

But thinking about it now yeah you're probably right that any wildcard headers are going to mean iterating through all the headers anyway (except in the edge case of someone declaring the entire set of "standard" http headers and then having a couple additional ones or some such thing.) I'll update it to use a boolean "has any wildcard headers" value (and add docs and tests)

sonnens commented:
@neuronull Updated this PR, I replaced the extra field with an enum & a function to build a vector of it from a Vec<String>. I think that should address the concerns about resource cost.

I also added a couple lines to the docs & the headers test, although I cannot for the life of me find the docs on how to actually *run* the test suite

neuronull commented on src/sources/http_server.rs:
```diff
@@ -1012,6 +1059,7 @@ mod tests {
             assert_eq!(log["key1"], "value1".into());
             assert_eq!(log["\"User-Agent\""], "test_client".into());
             assert_eq!(log["\"Upgrade-Insecure-Requests\""], "false".into());
+            assert_eq!(log["X-Test-Header"], "true".into());
```
This can be run locally by

```
cargo test --no-default-features -F sources-http_server sources::http_server
```

neuronull commented on src/sources/http_server.rs:
```diff
@@ -977,11 +1022,13 @@ mod tests {
             let mut headers = HeaderMap::new();
             headers.insert("User-Agent", "test_client".parse().unwrap());
             headers.insert("Upgrade-Insecure-Requests", "false".parse().unwrap());
+            headers.insert("X-Test-Header", "true".parse().unwrap());
```
It'd be useful to have a separate unit test case (in addition to this) - where that tests the "*" results in all the headers being added.

neuronull commented:
Thanks for iterating on the solution @sonnens 

> although I cannot for the life of me find the docs on how to actually run the test suite

The most direct way is 

```
cargo test --no-default-features -F sources-http_server sources::http_server

```
, it looks like the modified test case is failing.

You'll also want to run

```
make check-fmt
make check-clippy
```

The format lints can be fixed with `cargo-fmt`.

sonnens commented:
@neuronull here you go. This should take care of those concerns. It'll bail on error now if glob::pattern fails , made the docs changes you requested, added a test for matching everything & fixed the tests which work now. In doing so, I had to double check with RFC2616 that header keys are case *in*sensitive, but header values are sensitive, so I added a test for that as well.

neuronull commented:
Ah, I forgot to mention- this change since it is modifying the config options for the source, requires that we regenerate the docs for the component. You can do this by running

`make generate-component-docs` 

, and committing the resulting changes.

neuronull commented:
🙏  thanks for making the requested changes and overall iterating on the solution @sonnens !

urseberry commented on src/sources/http_server.rs:
```diff
@@ -88,10 +88,16 @@ pub struct SimpleHttpConfig {
 
     /// A list of HTTP headers to include in the log event.
     ///
+    /// Accepts wildcard (`*`) character for headers matching a specified pattern.
```
```suggestion
    /// Accepts the wildcard (`*`) character for headers matching a specified pattern.
```

neuronull replied:
addressed in https://github.com/vectordotdev/vector/pull/18922/commits/4c319da8cbdb8199be9ddeabc8ff6f253134899f

urseberry reviewed (approved):
One non-blocking grammar nit.

neuronull reviewed (approved):
We had just some small edit to the user facing docs and I added some code comments but otherwise this looks good to me.

neuronull commented:
CI in the merge queue failed due to a too many requests error returned from docker.

Take 2

neuronull commented:
This was kicked out of the merge queue due to a known issue that we hope to resolve soon- there was a performance 

_improvement_

 on one of the performance regression test cases. This incorrectly was perceived as a failure.

neuronull commented:
Thanks for the contribution @sonnens  !
